src/plugins/mf_priority.cpp

```diff
@@ -198,12 +198,181 @@ int check_queue_factor (flux_plugin_t *p,
+        json_t *jobid = json_integer (j);
+
+        if (!jobid)
+            goto error;
+
+        if (json_array_append (held_jobs, jobid) < 0) {
+            json_decref (jobid);
+            goto error;
+        }
+    }
+
+    return held_jobs;
+error:
+    json_decref (held_jobs);
+    return NULL;
+}
+
+
+/*
+ * Create a JSON object for a bank that a user belongs to.
+ */
+static json_t *pack_bank_info_object (
+                            const std::pair<std::string, struct bank_info> &b)
+{
+    json_t *bank_info, *held_jobs = NULL;
+
+    bank_info = json_object ();
+    if (!bank_info)
+        goto error;
+
+    held_jobs = add_held_jobs (b);
+    if (held_jobs == NULL)
+        goto error;
+
+    if (!(bank_info = json_pack ("{s:s, s:f, s:i, s:i, s:i, s:i, s:o, s:i}",
+                                 "bank", b.first.c_str (),
+                                 "fairshare", b.second.fairshare,
+                                 "max_run_jobs", b.second.max_run_jobs,
+                                 "cur_run_jobs", b.second.cur_run_jobs,
+                                 "max_active_jobs", b.second.max_active_jobs,
+                                 "cur_active_jobs", b.second.cur_active_jobs,
+                                 "held_jobs", held_jobs,
+                                 "active", b.second.active))) {
+            goto error;
+    }
+
+    return bank_info;
+error:
+    json_decref (held_jobs);
+    json_decref (bank_info);
+    return NULL;
+}
+
+
+/*
+ * For each bank that a user belongs to, create a JSON object for each bank and
+ * add it to the user JSON object.
+ */
+static json_t *banks_to_json (
+                    flux_plugin_t *p,
+                    std::pair<int, std::map<std::string, struct bank_info>> &u)
+{
+    json_t *bank_info, *banks = NULL;
+
+    banks = json_array (); // array of banks that user belongs to
+    if (!banks)
+        goto error;
+
+    for (auto const &b : u.second) {
+        if (!(bank_info = json_object ())) // JSON object for one bank
+            goto error;
+
+        bank_info = pack_bank_info_object (b);
+
+        if (json_array_append (banks, bank_info) < 0)
+            goto error;
+        
+        json_decref (bank_info);
+    }
+
+    return banks;
+error:
+    json_decref (banks);
+    json_decref (bank_info);
+    return NULL;
+}
+
+
+/*
+ * Iterate thrpugh each user in users map and create a JSON object for each
+ * user.
+ */
+static json_t *user_to_json (
+                    flux_plugin_t *p,
+                    std::pair<int, std::map<std::string, struct bank_info>> u)
+{
+    json_t *user = json_object (); // JSON object for one user
+    json_t *userid = NULL;
+
+    if (!user)
+        return NULL;
+
+    userid = json_integer (u.first);
+    if (!userid)
+        return NULL;
+
+    if (json_object_set_new (user, "userid", userid) < 0)
+        goto error;
+
+    if (json_object_set_new (user, "banks", banks_to_json (p, u)) < 0)
```

Comment:
Need to check for NULL return from `banks_to_json()`.

Also if the `json_object_set_new()` on "banks" fails, now userid is owned by user, so you will get a double free on the error path (userid is freed once when user is freed, then again explicitly).

Reply:
Ah, okay. Maybe I should just declare a separate JSON object to put the result of `banks_to_json ()` into. Then I can check for `NULL` before adding it with `json_object_set_new ()`.

> Also if the `json_object_set_new()` on "banks" fails, now userid is owned by user, so you will get a double free on the error path (userid is freed once when user is freed, then again explicitly).

I think I see what you're saying. So in this case, I do not need to explicitly free `userid` in the error path since it is already owned by `user`?

```diff
error:
    json_decref (user);
-   json_decref (userid);
    return NULL;
```

Reply:
Yes but you will need to free it if `json_object_set_new()` on `userid` fails.